### PR TITLE
[expo-splash-screen][ios] prefer AppDelegate's window over current keyWindow

### PR DESCRIPTION
**Pull request by esamelson**
# Why

resolves ENG-1373

**tl;dr this fixes the splash-screen/dev-menu interaction by changing the splash screen module's assumption that the splash screen will always be in the topmost/key UIWindow to an assumption that the splash screen will always be in the AppDelegate's UIWindow.**

The current splash screen module code on iOS gets a reference to a view controller (via UMCore) by assuming that the topmost/key UIWindow displayed on the screen at the moment `hideAsync` is called always contains the app's main view controller.

This assumption breaks down when the dev menu comes into play, because it's displayed in its own UIWindow which becomes the keyWindow when the dev menu is rendered. If `SplashScreen.hideAsync()` is called when the dev menu is open, we get the familiar `No native splash screen registered for given view controller. Call 'SplashScreen.show' for given view controller first.` warning and the splash screen never goes away, forcing the user to reload the app.

Unfortunately this can happen in what might be a fairly common scenario -- `expo init (tabs) && yarn add expo-dev-client && expo run:ios` since we automatically open the dev menu the first time the dev client is opened.

This PR changes the way that the splash screen module on iOS gets a reference to a view controller. As before, we first look for the Expo Go/`expo build`-specific `Util` module that provides the view controller in the managed workflow case, but then rather than falling back to the application's `keyWindow`, we first check try to read the AppDelegate's `window` property (where the main RCTRootViewController is stored by default in RN apps) before falling back to the `keyWindow`.

This shouldn't have any implications for managed apps or 99% or bare apps, but may introduce issues for any apps that are using React Native in an unconventional way -- i.e. the root AppDelegate's UIWindow doesn't contain the RCTRootView. I think it is pretty unlikely that any such apps will be making use of the `expo-splash-screen` module, and I think this solution is more correct than the current assumption, but open to discussion!

# How

- Implement EXSplashScreenModule's own `currentViewController` method, which reads from the managed workflow internal module if it exists (stealing this code from UMUtilities`) and then adds an intermediate check for the AppDelegate's UIWindow before falling back to the original `UMUtilities` code if such a UIWindow doesn't exist.

# Test Plan

```
expo init (tabs)
yarn add expo-dev-client
expo prebuild
```
- Make this change in local node_modules
- Build and run project from Xcode, then stop it without interacting at all with the dev menu
- `expo start --dev-client` then press `i`

Project should open and splash screen should hide while dev menu also appears automatically.

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by brentvatne — APPROVED** [review]
> may introduce issues for any apps that are using React Native in an unconventional way -- i.e. the root AppDelegate's UIWindow doesn't contain the RCTRootView. I think it is pretty unlikely that any such apps will be making use of the expo-splash-screen module, and I think this solution is more correct than the current assumption, but open to discussion!

what would the behavior be in this case? what do you think would be the most desirable?

**Comment by esamelson** [comment]
> what would the behavior be in this case? what do you think would be the most desirable?

The behavior in this case would be analogous to what happens currently when the dev menu is open -- the splash screen methods wouldn't work and would show a warning when called anytime the AppDelegate's UIWindow doesn't contain expo-splash-screen's splash screen view.

The best thing would be if the RN/unimodule infrastructure provided us with a reference to the root RN view controller -- then we wouldn't have to guess at all. But AFAIK this isn't available currently (would be happy to be wrong!).

**Comment by esamelson** [comment]
Closing this, going to try just iterating through all the application's windows instead which should hopefully take care of both of these cases.
